File: webflux-dialogue/src/main/java/com/study/webflux/rag/infrastructure/monitoring/document/PerformanceMetricsDocument.java
package com.study.webflux.rag.infrastructure.monitoring.document;

import java.time.Instant;
import java.util.List;
import java.util.Map;
import java.util.stream.Collectors;

import org.springframework.data.annotation.Id;
import org.springframework.data.mongodb.core.index.Indexed;
import org.springframework.data.mongodb.core.mapping.Document;

import com.study.webflux.rag.domain.monitoring.model.PerformanceMetrics;

@Document(collection = "performance_metrics")
public record PerformanceMetricsDocument(
	@Id String pipelineId,
	String status,
	@Indexed Instant startedAt,
	Instant finishedAt,
	@Indexed long totalDurationMillis,
	Long firstResponseLatencyMillis,
	Long lastResponseLatencyMillis,
	List<StagePerformanceDoc> stages,
	Map<String, Object> systemAttributes
) {
	public record StagePerformanceDoc(
		String stageName,
		String status,
		Instant startedAt,
		Instant finishedAt,
		long durationMillis,
		Map<String, Object> attributes) {
		public static StagePerformanceDoc fromDomain(PerformanceMetrics.StagePerformance domain) {
			return new StagePerformanceDoc(
				domain.stageName(),
				domain.status(),
				domain.startedAt(),
				domain.finishedAt(),
				domain.durationMillis(),
				sanitizeMapKeys(domain.attributes()));
		}

		public PerformanceMetrics.StagePerformance toDomain() {
			return new PerformanceMetrics.StagePerformance(
				stageName,
				status,
				startedAt,
				finishedAt,
				durationMillis,
				restoreMapKeys(attributes));
		}
	}

	public static PerformanceMetricsDocument fromDomain(PerformanceMetrics domain) {
		return new PerformanceMetricsDocument(
			domain.pipelineId(),
			domain.status(),
			domain.startedAt(),
			domain.finishedAt(),
			domain.totalDurationMillis(),
			domain.firstResponseLatencyMillis(),
			domain.lastResponseLatencyMillis(),
			domain.stages() == null
				? java.util.List.of()
				: domain.stages().stream().map(StagePerformanceDoc::fromDomain).toList(),
			sanitizeMapKeys(domain.systemAttributes()));
	}

	public PerformanceMetrics toDomain() {
		return new PerformanceMetrics(
			pipelineId,
			status,
			startedAt,
			finishedAt,
			totalDurationMillis,
			firstResponseLatencyMillis,
			lastResponseLatencyMillis,
			stages == null
				? java.util.List.of()
				: stages.stream().map(StagePerformanceDoc::toDomain).toList(),
			restoreMapKeys(systemAttributes));
	}

	private static Map<String, Object> sanitizeMapKeys(Map<String, Object> map) {
		if (map == null) {
			return null;
		}
		return map.entrySet().stream()
			.collect(Collectors.toMap(
				e -> e.getKey().replace(".", "__DOT__"),
				Map.Entry::getValue));
	}

	private static Map<String, Object> restoreMapKeys(Map<String, Object> map) {
		if (map == null) {
			return null;
		}
		return map.entrySet().stream()
			.collect(Collectors.toMap(
				e -> e.getKey().replace("__DOT__", "."),
				Map.Entry::getValue));
	}
}

File: app-miyou/src/main/kotlin/com/miyou/app/infrastructure/monitoring/document/PerformanceMetricsDocument.kt
package com.miyou.app.infrastructure.monitoring.document

import com.miyou.app.domain.monitoring.model.PerformanceMetrics
import org.springframework.data.annotation.Id
import org.springframework.data.mongodb.core.index.Indexed
import org.springframework.data.mongodb.core.mapping.Document
import java.time.Instant

@Document(collection = "performance_metrics")
data class PerformanceMetricsDocument(
    @Id val pipelineId: String,
    val status: String,
    @Indexed val startedAt: Instant?,
    val finishedAt: Instant?,
    @Indexed val totalDurationMillis: Long,
    val firstResponseLatencyMillis: Long?,
    val lastResponseLatencyMillis: Long?,
    val stages: List<StagePerformanceDoc>,
    val systemAttributes: Map<String, Any?>,
) {
    companion object {
        fun fromDomain(domain: PerformanceMetrics): PerformanceMetricsDocument =
            PerformanceMetricsDocument(
                pipelineId = domain.pipelineId,
                status = domain.status,
                startedAt = domain.startedAt,
                finishedAt = domain.finishedAt,
                totalDurationMillis = domain.totalDurationMillis,
                firstResponseLatencyMillis = domain.firstResponseLatencyMillis,
                lastResponseLatencyMillis = domain.lastResponseLatencyMillis,
                stages = domain.stages.map(StagePerformanceDoc::fromDomain),
                systemAttributes = sanitizeMapKeys(domain.systemAttributes),
            )
    }

    fun toDomain(): PerformanceMetrics =
        PerformanceMetrics(
            pipelineId,
            status,
            startedAt,
            finishedAt,
            totalDurationMillis,
            firstResponseLatencyMillis,
            lastResponseLatencyMillis,
            stages.map(StagePerformanceDoc::toDomain),
            restoreMapKeys(systemAttributes),
        )

    @Document
    data class StagePerformanceDoc(
        val stageName: String,
        val status: String,
        val startedAt: Instant?,
        val finishedAt: Instant?,
        val durationMillis: Long,
        val attributes: Map<String, Any?>,
    ) {
        companion object {
            fun fromDomain(domain: PerformanceMetrics.StagePerformance): StagePerformanceDoc =
                StagePerformanceDoc(
                    stageName = domain.stageName,
                    status = domain.status,
                    startedAt = domain.startedAt,
                    finishedAt = domain.finishedAt,
                    durationMillis = domain.durationMillis,
                    attributes = sanitizeMapKeys(domain.attributes),
                )
        }

        fun toDomain(): PerformanceMetrics.StagePerformance =
            PerformanceMetrics.StagePerformance(
                stageName = stageName,
                status = status,
                startedAt = startedAt,
                finishedAt = finishedAt,
                durationMillis = durationMillis,
                attributes = restoreMapKeys(attributes),
            )
    }
}

private fun sanitizeMapKeys(map: Map<String, Any?>): Map<String, Any?> =
    map.entries.associate { (key, value) ->
        encodeKey(key) to value
    }

private fun restoreMapKeys(map: Map<String, Any?>): Map<String, Any?> =
    map.entries.associate { (key, value) ->
        decodeKey(key) to value
    }

private fun encodeKey(key: String): String = key.replace("%", "%25").replace(".", "%2E")

private fun decodeKey(key: String): String = key.replace("%2E", ".").replace("%25", "%")
